File: backends/vapor/server/Sources/App/Web/Envelope.swift
import Vapor

/// 统一响应外壳 `{ success, data, message, errorCode }`。
/// 成功响应的 message/errorCode 必须显式输出 null（契约 strict equal 断言）。
enum Envelope {
    static func ok(_ data: JSON, status: HTTPStatus = .ok) -> Response {
        json(.object([
            "success": .bool(true),
            "data": data,
            "message": .null,
            "errorCode": .null,
        ]), status: status)
    }

    static func error(
        _ status: HTTPStatus,
        code: String,
        message: String,
        details: [(field: String, message: String)]?
    ) -> Response {
        var body: [String: JSON] = [
            "success": .bool(false),
            "data": .null,
            "message": .string(message),
            "errorCode": .string(code),
        ]
        if let details, !details.isEmpty {
            body["details"] = .array(details.map {
                .object(["field": .string($0.field), "message": .string($0.message)])
            })
        }
        return json(.object(body), status: status)
    }

    static func noContent() -> Response {
        Response(status: .noContent)
    }

    private static func json(_ value: JSON, status: HTTPStatus) -> Response {
        let data = (try? JSONEncoder().encode(value)) ?? Data("{}".utf8)
        var headers = HTTPHeaders()
        headers.contentType = .json
        return Response(status: status, headers: headers, body: .init(data: data))
    }
}

File: backends/axum/src/web/envelope.rs
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde_json::{json, Value};

/// 统一响应外壳 `{ success, data, message, errorCode }`。
/// serde_json::Value 的 Null 序列化为显式 null，天然满足契约 strict equal 断言。

pub fn ok(data: Value) -> Response {
    ok_with(StatusCode::OK, data)
}

pub fn ok_with(status: StatusCode, data: Value) -> Response {
    let body = json!({
        "success": true,
        "data": data,
        "message": null,
        "errorCode": null,
    });
    (status, Json(body)).into_response()
}

pub fn error(
    status: StatusCode,
    code: &str,
    message: &str,
    details: Option<&[(String, String)]>,
) -> Response {
    let mut body = json!({
        "success": false,
        "data": null,
        "message": message,
        "errorCode": code,
    });
    if let Some(details) = details {
        if !details.is_empty() {
            body["details"] = Value::Array(
                details
                    .iter()
                    .map(|(field, message)| json!({"field": field, "message": message}))
                    .collect(),
            );
        }
    }
    (status, Json(body)).into_response()
}

pub fn no_content() -> Response {
    StatusCode::NO_CONTENT.into_response()
}
